python/perf-kernels/flash-attention.py

```diff
@@ -159,53 +210,52 @@ def _attn_fwd(
     stride_vz, stride_vh, stride_vk, stride_vn,
     stride_oz, stride_oh, stride_om, stride_on,
     stride_bz, stride_bh, stride_bm, stride_bn,
-    Z, H,
-    N_CTX,
-    BLOCK_DMODEL: tl.constexpr,
+    H,
+    seqlen_q,
+    seqlen_k,
+    dropout_p,
+    philox_seed,
+    philox_offset_base,
+    encoded_softmax,
     STAGE: tl.constexpr,
     BLOCK_M: tl.constexpr,
+    BLOCK_DMODEL: tl.constexpr,
     BLOCK_N: tl.constexpr,
     pre_load_v: tl.constexpr,
     need_padding: tl.constexpr,
     extra_tokens_n: tl.constexpr,
-    bias_type: tl.constexpr
+    bias_type: tl.constexpr,
+    ENABLE_DROPOUT: tl.constexpr,
+    RETURN_ENCODED_SOFTMAX: tl.constexpr
 ):
     start_m = tl.program_id(0)
     off_hz = tl.program_id(1)
-    qvk_offset = off_hz * stride_qh
-
-    # block pointers
+    q_offset = off_hz * stride_qh
     Q_block_ptr = tl.make_block_ptr(
-        base=Q + qvk_offset,
-        shape=(N_CTX, BLOCK_DMODEL),
+        base=Q + q_offset,
+        shape=(seqlen_q, BLOCK_DMODEL),
         strides=(stride_qm, stride_qk),
         offsets=(start_m * BLOCK_M, 0),
         block_shape=(BLOCK_M, BLOCK_DMODEL),
-        order=(1, 0),
-    )
-    V_block_ptr = tl.make_block_ptr(
-        base=V + qvk_offset,
-        shape=(N_CTX, BLOCK_DMODEL),
-        strides=(stride_vk, stride_vn),
-        offsets=(0, 0),
-        block_shape=(BLOCK_N, BLOCK_DMODEL),
-        order=(1, 0),
+        order=(1, 0)
     )
+    k_offset = off_hz * stride_kh
     K_block_ptr = tl.make_block_ptr(
-        base=K + qvk_offset,
-        shape=(BLOCK_DMODEL, N_CTX),
+        base=K + k_offset,
+        shape=(BLOCK_DMODEL, seqlen_k),
         strides=(stride_kk, stride_kn),
         offsets=(0, 0),
         block_shape=(BLOCK_DMODEL, BLOCK_N),
-        order=(0, 1),
+        order=(0, 1)
     )
-    O_block_ptr = tl.make_block_ptr(
-        base=Out + qvk_offset,
-        shape=(N_CTX, BLOCK_DMODEL),
-        strides=(stride_om, stride_on),
-        offsets=(start_m * BLOCK_M, 0),
-        block_shape=(BLOCK_M, BLOCK_DMODEL),
-        order=(1, 0),
+    v_offset = off_hz * stride_vh
```

Comment:
kv always has the same shape. It would be preferable to keep this as a single term, so that it can consolidate the VGPRs when computing the pointers. Q is outside the loop but this is inside the loop so using as few VGPRs as possible is helpful.

Reply:
One problem is having the same shape does not mean having the same strides.

Reply:
That is a fair point.